### PR TITLE
[#11706] Allow overriding of values in `build.properties` in dev server

### LINKED ISSUE #11490: Support custom value for base system URL
Currently, the code assumes that the system is accessible at `https://<project-id>.appspot.com`. We need to modify the code such that the back-end can accept a customizable value for the base system URL.

Note that this is NOT an infrastructure work to map a custom domain name to the App Engine instance. This is just to support using the correct URL in emails and for redirects, and handle CORS if necessary.

### LINKED ISSUE #11706: Allow overriding of values in `build.properties` in dev server
Current: `build.properties` is a one-stop configuration file for both production and development instances.

Problem:
While this works, this presents a great inconvenience for developers who have to frequently deploy to production server, as they need to change some values back-and-forth depending on whether they are deploying or running local server. While it is true that the need for developers to use production server has been continuously diminishing, it is still not zero and will never be.

Solution:
- Maintain a dedicated dev server configuration file, e.g. `build-dev.properties`.
- When `Config.java` is initialized, the value of a property is by default read from `build.properties`. However, if the environment is determined to be dev server and the value exists in `build-dev.properties` (which also may not exist), the value will be read from `build-dev.properties` instead.
- Some properties such as `app.localdatastore.port`, `app.taskqueue.active`, and `app.enable.devserver.login` do not make sense in production server. We can optionally introduce another layer of logic here to just use pre-defined values for these properties in production server.
  - So is `app.frontenddev.url`, but this particular property can actually be re-purposed to concurrently solve #11490. May need a rename, though.

This solution draws inspiration from Spring Boot's `application.properties` and `application-*.properties`, for those who are familiar.

### PR DESCRIPTION
**Pull request by fsgmhoward**
Fixes #11706 
Fixes #11490 

Also, this PR supersedes and therefore closes #11531.

Idea for the fix is described in the issue. This is still work-in-progress.

### CODE REVIEW

File: src/main/java/teammates/common/util/Config.java [hunk 1]
```diff
@@ -103,53 +106,89 @@ public final class Config {
...
+        ENABLE_DATASTORE_BACKUP = Boolean.parseBoolean(
+                getProperty(properties, devProperties, "app.enable.datastore.backup", "false"));
+        MAINTENANCE = Boolean.parseBoolean(getProperty(properties, devProperties, "app.maintenance", "false"));
     }
 
     private Config() {
         // access static fields directly
     }
 
-    static String getBaseAppUrl() {
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/util/Config.java:151]: I don't think this method should be removed

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]
Aware that it's still ongoing, but there are some parts that are quite clearly not done the optimal way

File: src/main/java/teammates/common/util/Config.java [hunk 2]
```diff
@@ -212,21 +267,7 @@ public static boolean isDevServerLoginEnabled() {
      * {@code relativeUrl} must start with a "/".
      */
     public static AppUrl getFrontEndAppUrl(String relativeUrl) {
-        if (Config.isDevServer() && APP_FRONTENDDEV_URL != null) {
-            return new AppUrl(APP_FRONTENDDEV_URL + relativeUrl);
-        }
-
-        // In production, the back-end and front-end lives under the same domain
-        return getBackEndAppUrl(relativeUrl);
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/util/Config.java:220]: I don't think this behavior should be removed (not the scope of this issue anyway). You're likely right that `getBackEndAppUrl` can be removed, but what should be done here is replacing this with `new AppUrl(getBaseAppUrl() + relativeUrl)` directly.
Remember that `APP_FRONTEND_URL` is an optional property.

File: src/main/java/teammates/common/util/Config.java [hunk 3]
```diff
@@ -13,6 +13,10 @@
  */
 public final class Config {
 
+    // Normally, the following property values are loaded from build.properties file.
+    // When in dev environment and they appear in build-dev.properties,
+    // they are overridden by the values in build-dev.properties.
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/util/Config.java:18]: This comment is not in the right place. The class doesn't really care how the fields are set. This should be placed in the `static { ... }` block where all the fields are set.

File: src/main/java/teammates/common/util/Config.java [hunk 4]
```diff
@@ -103,53 +98,113 @@ public final class Config {
     /** The value of the "app.maintenance" in build.properties file. */
     public static final boolean MAINTENANCE;
 
+    // The following properties are not used in production server.
+    // So they will only be read from build-dev.properties file.
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/util/Config.java:102]: Similarly for this. With this said, the reordering of the properties may not be necessary.

File: src/main/java/teammates/common/util/Config.java [hunk 5]
```diff
@@ -103,53 +98,113 @@ public final class Config {
...
+        SEARCH_SERVICE_HOST = getProperty(properties, devProperties, "app.search.service.host");
+        ENABLE_DATASTORE_BACKUP = Boolean.parseBoolean(
+                getProperty(properties, devProperties, "app.enable.datastore.backup", "false"));
+        MAINTENANCE = Boolean.parseBoolean(getProperty(properties, devProperties, "app.maintenance", "false"));
+
+        // Dev property values, read directly from devProperties
+        APP_LOCALDATASTORE_PORT = Integer.parseInt(
+                devProperties.getProperty("app.localdatastore.port", "8484"));
+        ENABLE_DEVSERVER_LOGIN = Boolean.parseBoolean(
+                devProperties.getProperty("app.enable.devserver.login", "true"));
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/util/Config.java:184]: This particular property should now default to `false`. Although "only relevant in dev server", this one is actually used even for production server (see `isDevServerLoginEnabled()`), and in the event that `IS_DEV_SERVER` stops working properly, this is the last resort that stops the dev server login page from being displayed to public.

File: src/main/java/teammates/common/util/Config.java [hunk 6]
```diff
@@ -103,53 +98,113 @@ public final class Config {
...
     }
 
-    static String getBaseAppUrl() {
-        return isDevServer() ? "http://localhost:" + getPort() : "https://" + APP_ID + ".appspot.com";
+    private static String getProperty(Properties properties, Properties devProperties, String key, String defaultValue) {
+        if (IS_DEV_SERVER) {
+            return devProperties.getProperty(key, properties.getProperty(key, defaultValue));
+        } else {
+            return properties.getProperty(key, defaultValue);
+        }
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/util/Config.java:199]:
```java
if (IS_DEV_SERVER && devProperties.getProperty(key) != null) {
    return devProperties.getProperty(key);
}
return defaultValue == null ? properties.getProperty(key) : properties.getProperty(key, defaultValue);
```
Then the second method without `defaultValue` can just call this, no repetition needed.

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]
Implementation is good. Just need to improve on some details.

File: src/test/java/teammates/common/util/BuildPropertiesTest.java [hunk 7]
```diff
@@ -1,17 +0,0 @@
...
-import teammates.test.BaseTestCase;
-
-/**
- * SUT: {@link Config}.
- */
-public class BuildPropertiesTest extends BaseTestCase {
-
-    @Test
-    public void checkPresence() {
-        assertNotNull(Config.getBaseAppUrl());
```
[the remark this answers is not in the document]
Comment (fsgmhoward) [reply src/test/java/teammates/common/util/BuildPropertiesTest.java:14]: Thanks for pointing it out. It has been updated in the latest commit.

File: src/main/resources/build.template.properties [hunk 8]
```diff
@@ -19,17 +19,10 @@ app.region=us-central1
...
-app.localdatastore.port = 8484
-
-# This indicates whether task queues are active (e.g. items added to task queue will be queued for execution).
-# This flag is only used during development mode; in production, task queue will always be active.
-# In addition, during development mode, there is no "queueing", i.e. all tasks will be immediately executed.
-app.taskqueue.active=true
+# This is the URL for the front-end which the user is going to access.
+# This affects all URLs sent in outbound emails.
+# When using GAE, the default domain name is https://{app_id}.appspot.com
+app.frontend.url=https\://teammates-john.appspot.com
```
[the remark this answers is not in the document]
Comment (fsgmhoward) [reply src/main/resources/build.template.properties]: Thanks for pointing it out. It has been updated in the latest commit.

File: src/main/java/teammates/common/util/Config.java [hunk 9]
```diff
@@ -23,13 +23,7 @@ public final class Config {
     public static final String APP_VERSION;
 
     /** The value of the "app.frontenddev.url" in build.properties file. */
```
[the remark this answers is not in the document]
Comment (fsgmhoward) [reply src/main/java/teammates/common/util/Config.java]: Thanks for pointing it out. It has been updated in the latest commit.

File: src/main/resources/build.template.properties [hunk 10]
```diff
@@ -19,17 +19,12 @@ app.region=us-central1
...
-# This indicates whether task queues are active (e.g. items added to task queue will be queued for execution).
-# This flag is only used during development mode; in production, task queue will always be active.
-# In addition, during development mode, there is no "queueing", i.e. all tasks will be immediately executed.
-app.taskqueue.active=true
+# This is the URL for the front-end which the user is going to access.
+# This affects all URLs sent in outbound emails.
+# If not explicitly set:
+#   when using GAE, the default domain name is https://{app_id}.appspot.com
+#   when running locally, the default domain name is http://localhost:8080
+app.frontend.url=
```
Comment (wkurniawan07) [inline src/main/resources/build.template.properties:27]: I just remembered, an empty value for a property will not trigger the default value to be used. So for this particular property, the entire thing should be commented out instead:
```suggestion
# app.frontend.url=
```

File: src/main/java/teammates/common/util/Config.java [hunk 11]
```diff
@@ -103,53 +94,121 @@ public final class Config {
...
     private Config() {
         // access static fields directly
     }
 
-    static String getBaseAppUrl() {
-        return isDevServer() ? "http://localhost:" + getPort() : "https://" + APP_ID + ".appspot.com";
+    /**
+     * Returns the a default frontend URL if it is not set in property file(s).
+     */
+    public static String getDefaultFrontEndUrl() {
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/util/Config.java:184]: This can be package level visibility instead of public?

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]
Almost there!

File: src/main/java/teammates/common/util/Config.java (hunk 11 above)
Comment (fsgmhoward) [reply src/main/java/teammates/common/util/Config.java:184]: Oh yes. That makes sense. Thanks for pointing that out!

**Review by wkurniawan07 — APPROVED** [review]
LGTM, well done for this

**Comment by damithc** [comment]
@wkurniawan07 does this affect the production deployment process? e.g., any properties files need changing before I deploy the next version?

**Comment by wkurniawan07** [comment]
Not at all. This is all about development.
Although, to be safe, just make sure there is no property named `app.frontend.url`.

**Pull request merged**